src/adaXT/decision_tree/predict.pyx

```diff
@@ -0,0 +1,232 @@
+        for key in ht.keys():
+            indices = ht[key]
+            val = 1
+            count = len(indices)
+            if scale:
+                val = 1/count
+            matrix[np.ix_(indices, indices)] = val
+
+        return matrix
+
+
+cdef class PredictClassification(Predict):
+    def __cinit__(self, double[:, ::1] X, double[::1] Y, object root, **kwargs):
+        self.classes = np.unique(Y)
+
+    cdef int __find_max_index(self, double[::1] lst):
+        cdef:
+            int cur_max, i
+        cur_max = 0
+        for i in range(1, len(lst)):
+            if lst[cur_max] < lst[i]:
+                cur_max = i
+        return cur_max
+
+    def predict(self, object X, **kwargs):
+        cdef:
+            int i, cur_split_idx, idx, n_obs
+            double cur_threshold
+            object cur_node
+            double[:] Y
+        if not self.root:
+            raise AttributeError("The tree has not been fitted before trying to call predict")
+
+        # Make sure that x fits the dimensions.
+        X = Predict.__check_dimensions(self, X)
+        n_obs = X.shape[0]
+        Y = np.empty(n_obs)
+
+        for i in range(n_obs):
+            cur_node = self.root
+            while isinstance(cur_node, DecisionNode):
+                cur_split_idx = cur_node.split_idx
+                cur_threshold = cur_node.threshold
+                if X[i, cur_split_idx] < cur_threshold:
+                    cur_node = cur_node.left_child
+                else:
+                    cur_node = cur_node.right_child
+
+            idx = self.__find_max_index(cur_node.value)
+            if self.classes is not None:
+                Y[i] = self.classes[idx]
+        return Y
+
+    cpdef list predict_proba(self, object X):
+        cdef:
+            int i, cur_split_idx, n_obs
+            double cur_threshold
+            object cur_node
+            list ret_val
+
+        if not self.root:
+            raise AttributeError("The tree has not been fitted before trying to call predict_proba")
+
+        # Make sure that x fits the dimensions.
+        X = Predict.__check_dimensions(self, X)
+        n_obs = X.shape[0]
+        ret_val = []
+        for i in range(n_obs):
+            cur_node = self.root
+            while isinstance(cur_node, DecisionNode):
+                cur_split_idx = cur_node.split_idx
+                cur_threshold = cur_node.threshold
+                if X[i, cur_split_idx] < cur_threshold:
+                    cur_node = cur_node.left_child
+                else:
+                    cur_node = cur_node.right_child
+            if self.classes is not None:
+                ret_val.append(cur_node.value)
+        return ret_val
+
+
+cdef class PredictRegression(Predict):
+    def predict(self, object X, **kwargs):
+        cdef:
+            int i, cur_split_idx, n_obs
+            double cur_threshold
+            object cur_node
+            double[:] Y
+        if not self.root:
```

Comment:
Is this check a bit double given that its already done before initializing the class?